### PR TITLE
[ARTEMIS-550] fix up test to validate CORE and implement FQQN check b…

### PR DESCRIPTION
**Pull request by gtully**
…efore core auto createQueue call

### CODE REVIEW

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java [hunk 1]
```diff
@@ -698,8 +699,17 @@ private ActiveMQMessageConsumer createConsumer(final ActiveMQDestination dest,
...
             if (!response.isExists() || !response.getQueueNames().contains(dest.getSimpleAddress())) {
                if (response.isAutoCreateQueues()) {
+                  SimpleString queueNameToUse = dest.getSimpleAddress();
+                  SimpleString addressToUse = queueNameToUse;
+                  RoutingType routingTypeToUse = RoutingType.ANYCAST;
+                  if (CompositeAddress.isFullyQualified(queueNameToUse.toString())) {
+                     CompositeAddress compositeAddress = CompositeAddress.getQueueName(queueNameToUse.toString());
+                     addressToUse = new SimpleString(compositeAddress.getAddress());
+                     queueNameToUse = new SimpleString(compositeAddress.getQueueName());
+                     routingTypeToUse = RoutingType.MULTICAST;
```
Comment (michaelandrepearce) [inline artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: If this is a JMS Queue (des.isQueue), it should only be an ANYCAST

Comment (michaelandrepearce) [inline artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Is this related to Virtual Topics support for OpenWire?

Comment (gtully) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: it is, at least the use case in the test. It may be that some FQQN scenarios are not applicable to CORE.
I am not sure how to determine the routingTypeToUse if both the address and subscription queue need to be created.
Is there a way to know this is a fanout consumer queue/subscription queue rather than a simple ANYCAST address.
On the server in the openwire case there is a binding query such the address provides this information but I did not want to go down the route of another round trip.

Comment (gtully) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: The original test case that I added to, had both artemis and openwire jms clients in the mix, a parameterised test. It may be that the vt use case needs to be constrained to openwire because I think your expectation is correct. I guess in the case that the address already existed I may be able to find the routing type there. 
But when both address and queue need to be created core would be out of luck.

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: So if this is related to virtual topics why complicate things for the Core JMS client and introduce such problems it has JMS 2.0 spec as such shared durable subscribers, which is the spec equiv of such feature.

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Eg I think it should be constrained on core. As really virtual topics was a pre cursor custom impl of shared subscriptions in JMS 2 spec which we have already

Comment (graben) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Sorry, that's not correct vt and shared subscriptions are only similar but not equal. Shared subscriptions still need connection to have a clientid. Therefor those connections are bounded to a special use case. VTs has queue semantic. No "special" connections are needed. Such pooling of connections for almost everything sending/consuming/... is possible. Artemis still leaks such a feature on its core protocol.

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: No they don’t. Please see spec. Client ID is optional for a shared subscription.  Using a SharedDurableConsumer without setting a client id will give you queue like consuming semantics on a JMS Topic

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: This is there today.

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: https://docs.oracle.com/javaee/7/api/javax/jms/Session.html#createSharedDurableConsumer-javax.jms.Topic-java.lang.String-

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Please also see https://github.com/apache/activemq-artemis/blob/701ab1feba25d660b118dac79f77205326c9e90d/docs/migration-guide/en/VirtualTopics.md

Comment (graben) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Got it. I misinterpreted spec. I did a test and it works. Fine for me. Thx ;-)

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Great news!!!

Comment (graben) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: Well it depends. FQQN is a documented feature so usage this way might be possible. Make it work or document as not working.

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: And it does what it should that on a queue an anycast destination is created

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: I don’t see how that’s breaking, it’s been clear virtual topics as it is in activemq5 isn’t going to be ported as is, as use cases will be covered either with JMS 2.0 spec or FQQN (note FQQN != virtual topic). 

As noted here your use case is covered by jms shared subscriptions.

This PR atm breaks the documented address model to JMS mapping’s that exists. That a JMS queue maps onto Anycast Address and a JMS topic maps onto an Multicast Address.

Comment (mattrpav) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: @michaelandrepearce I shared some use cases and scenarios regarding 5.x virtual topics on 550. JMS 2.0 SharedDurable subscriptions is only one use case that virtual topics help with. However, there are several that JMS 2.0 SDS won’t cover that 5.x virtual topics do. Additionally, other brokers have these features, so I believe they are “messaging patterns” vs “ActiveMQ 5.x-only” thing. 

Check it out, and let me know if you want to hop on irc or a chat to go review it.  

At any rate, I think Artemis would benefit from supporting virtual topics (or whatever new name is more suitable)

Comment (michaelandrepearce) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: @mattrpav im on IRC. fyi im not saying all use cases are solved by SDS, but in this case it was. We should be looking to map use cases / functionality over, not lifting and shifting virt topics as is.

**Comment by michaelandrepearce** [comment]
If it helps so here is the conceptual logical model of VT:

vt

And here is the conceptual model of how JMS 2.0 SharedDurableSubscriber maps into Core Address model.

shared

If you note essentially the VT is the Address (its multicast), and then the queues are.....queues.

I don't see much difference now.

**Comment by mattrpav** [comment]
@michaelandrepearce the IBM MQ scenario I was referring to was the ability to use a named queue to back a topic subscription, not cloned subscriptions-- this provides the separation of producers from consumers in terms of "its a different destination". This enables better wild card management for permissions and other access features to destinations. ie.. topic://ORDERING.**  and queue://BILLING.**

I agree JMS 2.0 topic subscriptions solve the "same broker" and "same destination" use case for application-side ability to dynamically add additional consumers to a message flow without broker config or existing producer and consumer(s) impacts. 

Note: on the exclusive consumer part, it sounds like Artemis has a gap vs 5.x in that the exclusive consumer config is server-side only (separate issues raised as [ARTEMIS-853] and [ARTEMIS-856]).

The key feature in the broker-to-broker message scenarios is that the publishing destination is different than the consuming destination(s). In ActiveMQ 5.x, this allows for two important multi-broker messaging patterns, as well as enable the network of broker destination filtering (include / exclude), client permissions and destination policy mapping via wild card syntax sugar for administrators. See attached graphics

Multi-broker Use Case 1:
activemq-vt-multibroker-uc1

Multi-broker Use Case 2:
activemq-vt-multibroker-uc2

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java (hunk 1 above)
Comment (gtully) [reply artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java]: @michaelandrepearce  I agree with you, I will close out this PR. 
The discussion is the valuable bit.
For this use case, SDS is the answer. 
For the more general CORE FQQN scenario, using a prefix with a FQQN would be the other approach. It probably should support the prefix notation via the acceptor.
 ```multicast://VirtualTopic.Orders::Consumer.A```
This says, create me a sub queue called ```Consumer.A`` that is bound using ```multicast``` routing to address ```VirtualTopic.Orders```.

**Comment by gtully** [comment]
Hope we don't lose the pictures, but this PR as it stands is too simplistic and the use case is already covered by a shared durable sub.
the prefixed fqqn will be the way to go if we need to improve in this area.
https://activemq.apache.org/artemis/docs/2.3.0/address-model.html#configuring-a-prefix-to-connect-to-a-specific-routing-type

thanks for the feedback :-)

**Comment by michaelandrepearce** [comment]
@gtully agreed.

I think it be great if @mattrpav for the use cases we think we can't cover today, theyre detailed each in a JIRA, this way we can tackle one by one, or if they are possible today, then we can create Documents in how.

**Comment by mattrpav** [comment]
Will do
